### ISSUE TITLE
Driver => POS Communication

### ISSUE THREAD

**Issue by gohanman** [CORE-POS/IS4C#513]
TLDR: I dislike this and want to re-do it.

Communication from the scale (and now card terminal) driver to POS is ugly. It's slower than it needs to be, annoying to configure and debug, and fills the browser javascript console with tons of noise that makes pinpointing *other* javascript problems nearly impossible.

For the sake of clarity, here's what happens right now: when the driver needs to send a message to POS, it creates a uniquely named file in the ss-output directory containing the message. Meanwhile, POS is constantly polling via AJAX to see if files have been added to this directory.

Here's what should happen: POS should connect to a websocket or something similar to a websocket. When the driver needs to send a message, POS should receive it as a push notification via that websocket (or equivalent). This should be faster, if only slightly, than polling. It will eliminate all noise from constant AJAX requests in the javascript console. And POS should be able to give more useful, user-facing feedback if something goes wrong with the communication channel. The downside to re-doing this, of course, is that the current ugly system does *work* and hardware integration is very crucial functionality.

I see three potential avenues to approach this:

1. The driver includes a websocket server that POS can connect to. I've tried my hand at this in the past and it probably can work. The biggest upside is nothing gets added to the application stack for a lane. The downside is writing our own websocket server means re-inventing a wheel and potentially doing it poorly. I worry my implementation could have bugs or not be fully standards compliant. It would be unfortunate if a browser update broke scanner scale integration.

2. Install a dedicated message broker. My choice would be RabbitMQ having spent some time tinkering with it as a potential solution. The driver puts messages into the queue. POS subscribes to the queue. This is probably adds the most complication to installing the software although in the long run I think ease of debugging will be worthwhile. The advantage I see here is that other people have spent tons of time making a robust, reliable message queuing service and addressing the gotchas and side cases we don't know anything about.

3. Write our own message broker. Node.js makes it ridiculously simple to spin up a microservice for push notifications (and throwing node into the mix is of course trendy). Installation could be streamlined a bit since the out of the box configuration could be tailored exactly to CORE's usage (or there might not even need to be a configuration - the service just runs in "CORE mode" with no options to fiddle with). I don't think this increases potential for bugs nearly as much as option number one since node modules handle the vast, vast majority of server logic but any pitfalls relating to dropped connections or missed message deliveries would be on us to figure out rather than off loaded to a proven service like RabbitMQ.

I like option two the best, although there's some potential development fun with option number three.

I don't have any timeline in mind for this yet, but since I this would be a very significant change I wanted to publish the idea for comments/consideration/criticism/etc.

**Comment by joelbrock** [comment]
Disclosure:  This is an area of app development i only know superficially, so my understanding of the finer points is limited.  

I support moving in this direction as i DO understand the basic mechanics.  I think the polling method is truly one of the last great artifacts of IS4C.  I like the trendiness of Node.js, and wonder if there are other opportunities for employing Node in other aspects of hardware I/O.  (printing, KB input, ...) or server/lane processes.

RabbitMQ sounds like the easiest, but also possibly the best, solution of the 3.  What are the performance costs, if any?  Granted performance over the current polling method will in itself be significant.

I'm less concerned about the installation process than i used to be, since fewer people are out there installing CORE _withOUT_ our assistance.  So as long as I understand the installation/config process then i'm not bothered by further complicating installation.

My primary concern, and i know you are already sensitive to this but i'll state anyway, is backwards compatibility with existing production installs.  Basically we want to be 101% confident that installing any updates with newer dependencies interoperates flawlessly with legacy (CORE) systems.

Thanks as ever @gohanman for the great thought you continue to put into into driving CORE into a viable and vibrant future.  CHeers!  :+1:

**Comment by gohanman** [comment]
For the sake of clarification (hopefully): node wouldn't interact with hardware directly. I'm suggesting a 3rd service in between POS and the driver that strictly handles delivering messages from the driver to POS. So the C# driver would still be talking to the scale/terminal, but instead of writing out to a file it would submit the message to the node service and the node service would re-send the message as a push notification to any subscribed browser windows. The idea with RabbitMQ is basically the same with RabbitMQ in place of "node service".

I don't think there's any performance costs; if anything there should be performance gains. Polling certainly isn't the most efficient communication mechanism. My concerns, besides installation steps, are edge cases. If the driver submits something to the messaging service and the subscribed browser navigates to another page at the exact same time, give or take a couple milliseconds, is that message going to be lost? I'm expecting a thoroughly tested/studied/proven service like RabbitMQ is likely to handle this kind of thing better than something we throw together ourselves using node.js and minimal expertise (if any) in distributed data sources and consistency. On the other hand, those edge cases might be sufficiently rare that a custom service that "just works" without any configuration at all is a decent trade off.

Backwards compatibility is certainly an important consideration. I think it would actually be ideal to keep polling around and have both sides detect the message server and upgrade the connection. It would be even better if they could fallback to polling if the message server crashes.

**Comment by gohanman** [comment]
Ongoing process documentation:

### Installing RabbitMQ
This is presumably a simple package manage command on most Linux distros. In Windows, download the RabbitMQ binary installer. If necessary, it will prompt you to download and install erlang first. Next, enable websockets. Locate the rabbitmq-plugins application and run
```
rabbitmq-plugins enable rabbitmq_web_stomp
```
Then go to http://127.0.0.1:15674/stomp. It should be a page that says welcome to SockJS. Restarting the rabbitmq service may be necessary.

### Compile NewMagellan w/ RabbitMQ support
Get the .NET library (https://www.rabbitmq.com/dotnet.html). Create a directory named *rabbitmq* under NewMagellan and place the RabbitMQ.Client.dll file there. Recompile and restart the driver.

### Usage
There is no configuration. When POS starts up, it checks connectivity on port 15674. If that port is listening, it sends the driver a message to use RabbitMQ. Otherwise it sends the driver a message not to use RabbitMQ. Some back-and-forth handshaking might be cleaner, but at the moment don't run RabbitMQ on a lane unless the NewMagellan driver supports it.

### Linux support
No idea. The RabbitMQ .NET library claims to be Mono compatible, but I don't know if the downloadable DLL is cross platform or if it needs to be manually built from source using mono. I want to verify basic operations before digging into cross-platform issues.

**Comment by gohanman** [comment]
New documentation:
https://github.com/CORE-POS/IS4C/wiki/Hardware-Driver

NewMagellan remains one of the more black-magic components of the project. I sketched out a bit about how it interacts with POS. Plus a picture! A more concise summary of this thread could be: I want to replace the blue arrow.

**Comment by gohanman** [comment]
Moved this into documentation.
https://github.com/CORE-POS/IS4C/wiki/Using-RabbitMQ
